Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util/clientcache.go

```diff
@@ -21,17 +21,50 @@ import (
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
 )
 
-// clientCache caches previously loaded clients for reuse, and ensures MatchServerVersion
+// KubernetesClientCache is a cache of Kubernetes clients.
+type KubernetesClientCache struct {
+	*ClientCache
+}
+
+func NewKubernetesClientCache(loader clientcmd.ClientConfig, matchVersion bool, defaultFn ConfigDefaultsFunc) KubernetesClientCache {
+	newFn := func(config *client.Config) (interface{}, error) { return client.New(config) }
+	return KubernetesClientCache{NewClientCache(loader, matchVersion, defaultFn, newFn)}
+}
+
+func (c KubernetesClientCache) ClientForVersion(version string) (*client.Client, error) {
+	cli, err := c.ClientCache.ClientForVersion(version)
+	if err != nil {
+		return nil, err
+	}
+	return cli.(*client.Client), nil
+}
+
+// ClientCache caches previously loaded client objects for reuse, and ensures MatchServerVersion
 // is invoked only once
-type clientCache struct {
+type ClientCache struct {
 	loader        clientcmd.ClientConfig
-	clients       map[string]*client.Client
+	clients       map[string]interface{}
 	defaultConfig *client.Config
 	matchVersion  bool
+	setDefaultsFn ConfigDefaultsFunc
+	newClientFn   NewClientFunc
+}
+
+type ConfigDefaultsFunc func(*client.Config)
+type NewClientFunc func(*client.Config) (interface{}, error)
+
+func NewClientCache(loader clientcmd.ClientConfig, matchVersion bool, defaultFn ConfigDefaultsFunc, newFn NewClientFunc) *ClientCache {
```

Comment:
Describe when the default function is called.  Once per `defaultConfig` or once per `newFn`,